File: 2015/13/solution.js
let rules = {}

const addRule = (line) => {
  const matcher = /^(.*) would (lose|gain) (\d+) happiness units? by sitting next to (.*)\.$/
  let [ , person, action, happiness, neighbour ] = line.match(matcher)
  if (!rules[person]) rules[person] = {}
  rules[person][neighbour] = parseInt(happiness) * (action === 'lose' ? -1 : 1)
}

const permuteSeats = (people) => {
  if (people.length <= 1) return people
  let result = []
  for (let i = 0; i < people.length; i++) {
    const person = people[i]
    people.splice(i, 1)
    permuteSeats(people).forEach((permutation) => {
      result.push([ person ].concat(permutation))
    })
    people.splice(i, 0, person)
  }
  return result
}

const happiness = (person, neighbour) => (rules[person] || {})[neighbour] || 0

const pairHappiness = (person1, person2) => {
  return happiness(person1, person2) + happiness(person2, person1)
}

const permutationHappiness = (people) => {
  let happiness = 0
  for (let i = 0; i < people.length - 1; i++) {
    happiness += pairHappiness(people[i], people[i + 1])
  }
  happiness += pairHappiness(people[people.length - 1], people[0])
  return happiness
}

const solve = (additionalPeople = []) => {
  let maxHappiness = 0
  permuteSeats(Object.keys(rules).concat(additionalPeople)).forEach((permutation) => {
    const happiness = permutationHappiness(permutation)
    if (happiness > maxHappiness) {
      maxHappiness = happiness
    }
  })
  return maxHappiness
}

const report = () => {
  console.log('Part 1', solve())
  console.log('Part 2', solve([ 'Ian' ]))
}

require('readline').
    createInterface({ input: require('fs').createReadStream('input.txt') }).
    on('line', addRule).
    on('close', report)

File: 2015/13/solution.rb
def permute_seats(people)
  return [ people ] if people.length <= 1
  result = []
  people.each do |person|
    people.delete(person)
    permute_seats(people).each do |permutation|
      result.push([ person ] + permutation)
    end
    people.unshift(person)
  end
  result
end

def solve(rules, additionalPeople = [])
  max = 0
  permute_seats(rules.keys + additionalPeople).each do |people|
    happiness = 0
    for i in 0...(people.length - 1)
      happiness += rules[people[i]][people[i + 1]] + rules[people[i + 1]][people[i]]
    end
    happiness += rules[people.last][people.first] + rules[people.first][people.last]
    max = happiness if happiness > max
  end
  max
end

rules = Hash.new do |hash, key|
  hash[key] = Hash.new(0)
end

File.open('input.txt').each_line do |line|
  matcher = /^(.*) would (lose|gain) (\d+) happiness units? by sitting next to (.*)\.$/
  person, action, happiness, neighbour = matcher.match(line)[1..4]
  rules[person][neighbour] = happiness.to_i * (action === 'lose' ? -1 : 1)
end

puts 'Part 1', solve(rules)
puts 'Part 2', solve(rules, [ 'Ian' ])
